usr.man/man3/strftime.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the American National Standards Committee X3, on Information
.\" Processing Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)strftime.3	8.1 (Berkeley) 6/4/93
.\" $FreeBSD$
.\"
.TH "STRFTIME" "3" "June 25, 2012" "GNO" "Library Functions Manual"
.SH "NAME"
\fBstrftime\fR
\- format date and time
.SH "LIBRARY"
Standard C Library (libc, \-lc)
.SH "SYNOPSIS"
\fB#include <time.h>\fR
.sp
\fIsize_t\fR
\fBstrftime\fR(\fIchar\ *\ restrict\ buf\fR,
\fIsize_t\ maxsize\fR,
\fIconst\ char\ *\ restrict\ format\fR,
\fIconst\ struct\ tm\ *\ restrict\ timeptr\fR);
.PP
\fIsize_t\fR
\fBstrftime_l\fR(\fIchar\ *restrict\ buf\fR, \fIsize_t\ maxsize\fR, \fIconst\ char\ *\ restrict\ format\fR, \fIconst\ struct\ tm\ *restrict\ timeptr\fR, \fIlocale_t\ loc\fR);
.SH "DESCRIPTION"
The
\fBstrftime\fR()
function formats the information from
\fItimeptr\fR
into the buffer
\fIbuf\fR
according to the string pointed to by
\fIformat\fR.
The function
\fBstrftime_l\fR()
does the same as
\fBstrftime\fR()
but takes an explicit locale rather than using the current locale.
.PP
The
\fIformat\fR
string consists of zero or more conversion specifications and
ordinary characters.
All ordinary characters are copied directly into the buffer.
A conversion specification consists of a percent sign
```%'''
and one other character.
.PP
No more than
\fImaxsize\fR
characters will be placed into the array.
If the total number of resulting characters, including the terminating
NUL character, is not more than
\fImaxsize\fR,
\fBstrftime\fR()
returns the number of characters in the array, not counting the
terminating NUL.
Otherwise, zero is returned and the buffer contents are indeterminate.
.PP
The conversion specifications are copied to the buffer after expansion
as follows:-
.TP \fB\&%A\fR
is replaced by national representation of the full weekday name.
.TP \fB%a\fR
is replaced by national representation of
the abbreviated weekday name.
.TP \fB\&%B\fR
is replaced by national representation of the full month name.
.TP \fB%b\fR
is replaced by national representation of
the abbreviated month name.
.TP \fB\&%C\fR
is replaced by (year / 100) as decimal number; single
digits are preceded by a zero.
.TP \fB%c\fR
is replaced by national representation of time and date.
.TP \fB\&%D\fR
is equivalent to
``\fR%m/%d/%y\fR''.
.TP \fB%d\fR
is replaced by the day of the month as a decimal number (01-31).
.TP \fB%E* %O*\fR
POSIX locale extensions.
The sequences
%Ec %EC %Ex %EX %Ey %EY
%Od %Oe %OH %OI %Om %OM
%OS %Ou %OU %OV %Ow %OW %Oy
are supposed to provide alternate
representations.
.sp
Additionally %OB implemented
to represent alternative months names
(used standalone, without day mentioned).
.TP \fB%e\fR
is replaced by the day of the month as a decimal number (1-31); single
digits are preceded by a blank.
.TP \fB\&%F\fR
is equivalent to
``\fR%Y-%m-%d\fR''.
.TP \fB\&%G\fR
is replaced by a year as a decimal number with century.
This year is the one that contains the greater part of
the week (Monday as the first day of the week).
.TP \fB%g\fR
is replaced by the same year as in
``\fR%G\fR'',
but as a decimal number without century (00-99).
.TP \fB\&%H\fR
is replaced by the hour (24-hour clock) as a decimal number (00-23).
.TP \fB%h\fR
the same as
\fB%b\fR.
.TP \fB\&%I\fR
is replaced by the hour (12-hour clock) as a decimal number (01-12).
.TP \fB%j\fR
is replaced by the day of the year as a decimal number (001-366).
.TP \fB%k\fR
is replaced by the hour (24-hour clock) as a decimal number (0-23);
single digits are preceded by a blank.
.TP \fB%l\fR
is replaced by the hour (12-hour clock) as a decimal number (1-12);
single digits are preceded by a blank.
.TP \fB\&%M\fR
is replaced by the minute as a decimal number (00-59).
.TP \fB%m\fR
is replaced by the month as a decimal number (01-12).
.TP \fB%n\fR
is replaced by a newline.
.TP \fB%O*\fR
the same as
\fB%E*\fR.
.TP \fB%p\fR
is replaced by national representation of either
"ante meridiem" (a.m.)
or
"post meridiem" (p.m.)
as appropriate.
.TP \fB\&%R\fR
is equivalent to
``\fR%H:%M\fR''.
.TP \fB%r\fR
is equivalent to
``\fR%I:%M:%S %p\fR''.
.TP \fB\&%S\fR
is replaced by the second as a decimal number (00-60).
.TP \fB%s\fR
is replaced by the number of seconds since the Epoch, UTC (see
mktime(3)).
.TP \fB\&%T\fR
is equivalent to
``\fR%H:%M:%S\fR''.
.TP \fB%t\fR
is replaced by a tab.
.TP \fB\&%U\fR
is replaced by the week number of the year (Sunday as the first day of
the week) as a decimal number (00-53).
.TP \fB%u\fR
is replaced by the weekday (Monday as the first day of the week)
as a decimal number (1-7).
.TP \fB\&%V\fR
is replaced by the week number of the year (Monday as the first day of
the week) as a decimal number (01-53).
If the week containing January
1 has four or more days in the new year, then it is week 1; otherwise
it is the last week of the previous year, and the next week is week 1.
.TP \fB%v\fR
is equivalent to
``\fR%e-%b-%Y\fR''.
.TP \fB\&%W\fR
is replaced by the week number of the year (Monday as the first day of
the week) as a decimal number (00-53).
.TP \fB%w\fR
is replaced by the weekday (Sunday as the first day of the week)
as a decimal number (0-6).
.TP \fB\&%X\fR
is replaced by national representation of the time.
.TP \fB%x\fR
is replaced by national representation of the date.
.TP \fB\&%Y\fR
is replaced by the year with century as a decimal number.
.TP \fB%y\fR
is replaced by the year without century as a decimal number (00-99).
.TP \fB\&%Z\fR
is replaced by the time zone name.
.TP \fB%z\fR
is replaced by the time zone offset from UTC; a leading plus sign stands for
east of UTC, a minus sign for west of UTC, hours and minutes follow
with two digits each and no delimiter between them (common form for
RFC 822 date headers).
.TP \fB%+\fR
is replaced by national representation of the date and time
(the format is similar to that produced by
date(1)).
.TP \fB%-*\fR
GNU libc extension.
Do not do any padding when performing numerical outputs.
.TP \fB%_*\fR
GNU libc extension.
Explicitly specify space for padding.
.TP \fB%0*\fR
GNU libc extension.
Explicitly specify zero for padding.
.TP \fB%%\fR
is replaced by
`%'.
.SH "SEE ALSO"
date(1),
printf(1),
ctime(3),
printf(3),
strptime(3),
wcsftime(3)
.SH "STANDARDS"
The
\fBstrftime\fR()
function
conforms to
ISO/IEC 9899:1990 (``ISO C90'')
with a lot of extensions including
`\&%C',
`\&%D',
`%E*',
`%e',
`%G',
`%g',
`%h',
`%k',
`%l',
`%n',
`%O*',
`\&%R',
`%r',
`%s',
`\&%T',
`%t',
`%u',
`\&%V',
`%z',
`%+'.
.PP
The peculiar week number and year in the replacements of
`%G',
`%g'
and
`\&%V'
are defined in ISO 8601: 1988.
The
\fBstrftime_l\fR()
function conforms to
IEEE Std 1003.1-2008 (``POSIX.1'').
.SH "BUGS"
There is no conversion specification for the phase of the moon.
.PP
The
\fBstrftime\fR()
function does not correctly handle multibyte characters in the
\fIformat\fR
argument.


lib/libc/stdtime/strftime.c

/*
 * Copyright (c) 1989 The Regents of the University of California.
 * All rights reserved.
 *
 * Copyright (c) 2011 The FreeBSD Foundation
 * All rights reserved.
 * Portions of this software were developed by David Chisnall
 * under sponsorship from the FreeBSD Foundation.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley. The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 * $Id: strftime.c,v 1.2 2012/08/26 02:54:59 gdr Exp $
 *
 * This file is formatted for tab stops every 8 columns.
 */

#ifdef __ORCAC__
#define NOID
#endif

#ifdef LIBC_RCS
static const char rcsid[] =
	"$Id: strftime.c,v 1.2 2012/08/26 02:54:59 gdr Exp $";
#endif

#ifndef lint
#ifndef NOID
static const char	elsieid[] = "@(#)strftime.3	8.3";
/*
 * Based on the UCB version with the ID appearing below.
 * This is ANSIish only when "multibyte character == plain character".
 */
#endif /* !defined NOID */
#endif /* !defined lint */

#include "private.h"

#if defined(LIBC_SCCS) && !defined(lint)
static const char	sccsid[] = "@(#)strftime.c	5.4 (Berkeley) 3/14/89";
#endif /* LIBC_SCCS and not lint */

#include "tzfile.h"
#include "timelocal.h"
#ifdef __ORCAC__
#include <string.h>		/* for memcpy() */
#endif


static char *	_add (const char *, char *, const char *);
static char *	_conv (int, const char *, char *, const char *);
static char *	_fmt (const char *, const struct tm *, char *, const char *);
static char *   _secs (const struct tm *, char *, const char *);
static char *	_yconv(int, int, int, int, char *, const char *);

size_t strftime (char *, size_t, const char *, const struct tm *);

extern char *	tzname[];

#define	IN_NONE	0
#define	IN_SOME	1
#define	IN_THIS	2
#define	IN_ALL	3

#define	PAD_DEFAULT	0
#define	PAD_LESS	1
#define	PAD_SPACE	2
#define	PAD_ZERO	3

static const char fmt_padding[][4][5] = {
	/* DEFAULT,	LESS,	SPACE,	ZERO */
#define	PAD_FMT_MONTHDAY	0
#define	PAD_FMT_HMS		0
#define	PAD_FMT_CENTURY		0
#define	PAD_FMT_SHORTYEAR	0
#define	PAD_FMT_MONTH		0
#define	PAD_FMT_WEEKOFYEAR	0
#define	PAD_FMT_DAYOFMONTH	0
	{ "%02d",	"%d",	"%2d",	"%02d" },
#define	PAD_FMT_SDAYOFMONTH	1
#define	PAD_FMT_SHMS		1
	{ "%2d",	"%d",	"%2d",	"%02d" },
#define	PAD_FMT_DAYOFYEAR	2
	{ "%03d",	"%d",	"%3d",	"%03d" },
#define	PAD_FMT_YEAR		3
	{ "%04d",	"%d",	"%4d",	"%04d" }
};

#ifdef __ORCAC__
static void tzset(void) {}
static char *tzname[] = { NULL, NULL };
#endif

size_t
strftime(char *s, const size_t maxsize, const char *format,
	const struct tm *t)
{
	char *p;

	tzset();
	p = _fmt(((format == NULL) ? "%c" : format), t, s, s + maxsize);
	if (p == s + maxsize)
		return 0;
	*p = '\0';
	return p - s;
}

static char *
_fmt(const char *format, const struct tm *t, char *pt,
	const char *ptlim)
{
	int Ealternative, Oalternative, PadIndex;
#ifdef __ORCAC__
	#define tptr (&_C_time_locale)
#else
	struct lc_time_T *tptr = __get_current_time_locale(__get_locale());
#endif

	for ( ; *format; ++format) {
		if (*format == '%') {
			Ealternative = 0;
			Oalternative = 0;
			PadIndex	 = PAD_DEFAULT;
label:
			switch (*++format) {
			case '\0':
				--format;
				break;
			case 'A':
				pt = _add((t->tm_wday < 0 || t->tm_wday >= DAYSPERWEEK) ?
					"?" : tptr->weekday[t->tm_wday],
					pt, ptlim);
				continue;
			case 'a':
				pt = _add((t->tm_wday < 0 || t->tm_wday >= DAYSPERWEEK) ?
					"?" : tptr->wday[t->tm_wday],
					pt, ptlim);
				continue;
			case 'B':
				pt = _add((t->tm_mon < 0 ||
					t->tm_mon >= MONSPERYEAR) ?
					"?" : (Oalternative ? tptr->alt_month :
					tptr->month)[t->tm_mon],
					pt, ptlim);
				continue;

				continue;
			case 'b':
			case 'h':
				pt = _add((t->tm_mon < 0 || t->tm_mon >= MONSPERYEAR) ?
					"?" : tptr->mon[t->tm_mon],
					pt, ptlim);
				continue;
			case 'C':
				/*
				 * %C used to do a...
				 *	_fmt("%a %b %e %X %Y", t);
				 * ...whereas now POSIX 1003.2 calls for
				 * something completely different.
				 * (ado, 1993-05-24)
				 */
				pt = _yconv(t->tm_year, TM_YEAR_BASE, 1, 0,
					pt, ptlim);
				continue;
			case 'c':
				pt = _fmt(tptr->c_fmt, t, pt, ptlim);
				continue;
			case 'D':
				pt = _fmt("%m/%d/%y", t, pt, ptlim);
				continue;
			case 'd':
				pt = _conv(t->tm_mday,
					fmt_padding[PAD_FMT_DAYOFMONTH][PadIndex],
					pt, ptlim);
				continue;
			case 'E':
				if (Ealternative || Oalternative)
					break;
				Ealternative++;
				goto label;
			case 'O':
				/*
				 * C99 locale modifiers.
				 * The sequences
				 *	%Ec %EC %Ex %EX %Ey %EY
				 *	%Od %oe %OH %OI %Om %OM
				 *	%OS %Ou %OU %OV %Ow %OW %Oy
				 * are supposed to provide alternate
				 * representations.
				 *
				 * FreeBSD extension
				 *      %OB
				 */
				if (Ealternative || Oalternative)
					break;
				Oalternative++;
				goto label;
			case 'e':
				pt = _conv(t->tm_mday,
					fmt_padding[PAD_FMT_SDAYOFMONTH][PadIndex],
					pt, ptlim);
				continue;
			case 'F':
				pt = _fmt("%Y-%m-%d", t, pt, ptlim);
				continue;
			case 'H':
				pt = _conv(t->tm_hour, fmt_padding[PAD_FMT_HMS][PadIndex],
					pt, ptlim);
				continue;
			case 'I':
				pt = _conv((t->tm_hour % 12) ?
					(t->tm_hour % 12) : 12,
					fmt_padding[PAD_FMT_HMS][PadIndex],
					pt, ptlim);
				continue;
			case 'j':
				pt = _conv(t->tm_yday + 1,
					fmt_padding[PAD_FMT_DAYOFYEAR][PadIndex],
					pt, ptlim);
				continue;
			case 'k':
				/*
				 * This used to be...
				 *	_conv(t->tm_hour % 12 ?
				 *		t->tm_hour % 12 : 12, 2, ' ');
				 * ...and has been changed to the below to
				 * match SunOS 4.1.1 and Arnold Robbins'
				 * strftime version 3.0. That is, "%k" and
				 * "%l" have been swapped.
				 * (ado, 1993-05-24)
				 */
				 pt = _conv(t->tm_hour, fmt_padding[PAD_FMT_SHMS][PadIndex],
				 	pt, ptlim);
				continue;
#ifdef KITCHEN_SINK
			case 'K':
				/*
				** After all this time, still unclaimed!
				*/
				pt = _add("kitchen sink", pt, ptlim);
				continue;
#endif /* defined KITCHEN_SINK */
			case 'l':
				/*
				 * This used to be...
				 *	_conv(t->tm_hour, 2, ' ');
				 * ...and has been changed to the below to
				 * match SunOS 4.1.1 and Arnold Robbin's
				 * strftime version 3.0. That is, "%k" and
				 * "%l" have been swapped.
				 * (ado, 1993-05-24)
				 */
				pt = _conv((t->tm_hour % 12) ?
					(t->tm_hour % 12) : 12,
					fmt_padding[PAD_FMT_SHMS][PadIndex],
					pt, ptlim);
				continue;
			case 'M':
				pt = _conv(t->tm_min, fmt_padding[PAD_FMT_HMS][PadIndex],
					pt, ptlim);
				continue;
			case 'm':
				pt = _conv(t->tm_mon + 1,
					fmt_padding[PAD_FMT_MONTH][PadIndex],
					pt, ptlim);
				continue;
			case 'n':
				pt = _add("\n", pt, ptlim);
				continue;
			case 'p':
				pt = _add((t->tm_hour >= (HOURSPERDAY / 2)) ?
					tptr->pm : tptr->am,
					pt, ptlim);
				continue;
			case 'R':
				pt = _fmt("%H:%M", t, pt, ptlim);
				continue;
			case 'r':
				pt = _fmt(tptr->ampm_fmt, t, pt, ptlim);
				continue;
			case 'S':
				pt = _conv(t->tm_sec, fmt_padding[PAD_FMT_HMS][PadIndex],
					pt, ptlim);
				continue;
			case 's':
				pt = _secs(t, pt, ptlim);
				continue;
			case 'T':
				pt = _fmt("%H:%M:%S", t, pt, ptlim);
				continue;
			case 't':
				pt = _add("\t", pt, ptlim);
				continue;
			case 'U':
				pt = _conv((t->tm_yday + DAYSPERWEEK -
					t->tm_wday) / DAYSPERWEEK,
					fmt_padding[PAD_FMT_WEEKOFYEAR][PadIndex],
					pt, ptlim);
				continue;
			case 'u':
				/*
				 * From Arnold Robbins' strftime version 3.0:
				 * "ISO 8601: Weekday as a decimal number
				 * [1 (Monday) - 7]"
				 * (ado, 1993-05-24)
				 */
				pt = _conv((t->tm_wday == 0) ?
					DAYSPERWEEK : t->tm_wday,
					"%d", pt, ptlim);
				continue;
			case 'V':	/* ISO 8601 week number */
			case 'G':	/* ISO 8601 year (four digits) */
			case 'g':	/* ISO 8601 year (two digits) */
/*
 * From Arnold Robbins' strftime version 3.0: "the week number of the
 * year (the first Monday as the first day of week 1) as a decimal number
 * (01-53)."
 * (ado, 1993-05-24)
 *
 * From "http://www.ft.uni-erlangen.de/~mskuhn/iso-time.html" by Markus Kuhn:
 * "Week 01 of a year is per definition the first week which has the
 * Thursday in this year, which is equivalent to the week which contains
 * the fourth day of January. In other words, the first week of a new year
 * is the week which has the majority of its days in the new year. Week 01
 * might also contain days from the previous year and the week before week
 * 01 of a year is the last week (52 or 53) of the previous year even if
 * it contains days from the new year. A week starts with Monday (day 1)
 * and ends with Sunday (day 7). For example, the first week of the year
 * 1997 lasts from 1996-12-30 to 1997-01-05..."
 * (ado, 1996-01-02)
 */
				{
					int	year;
					int	base;
					int	yday;
					int	wday;
					int	w;

					year = t->tm_year;
					base = TM_YEAR_BASE;
					yday = t->tm_yday;
					wday = t->tm_wday;
					for ( ; ; ) {
						int	len;
						int	bot;
						int	top;

						len = isleap_sum(year, base) ?
							DAYSPERLYEAR :
							DAYSPERNYEAR;
						/*
						 * What yday (-3 ... 3) does
						 * the ISO year begin on?
						 */
						bot = ((yday + 11 - wday) %
							DAYSPERWEEK) - 3;
						/*
						 * What yday does the NEXT
						 * ISO year begin on?
						 */
						top = bot -
							(len % DAYSPERWEEK);
						if (top < -3)
							top += DAYSPERWEEK;
						top += len;
						if (yday >= top) {
							++base;
							w = 1;
							break;
						}
						if (yday >= bot) {
							w = 1 + ((yday - bot) /
								DAYSPERWEEK);
							break;
						}
						--base;
						yday += isleap_sum(year, base) ?
							DAYSPERLYEAR :
							DAYSPERNYEAR;
					}
#ifdef XPG4_1994_04_09
					if ((w == 52 &&
						t->tm_mon == TM_JANUARY) ||
						(w == 1 &&
						t->tm_mon == TM_DECEMBER))
							w = 53;
#endif /* defined XPG4_1994_04_09 */
					if (*format == 'V')
						pt = _conv(w, fmt_padding[PAD_FMT_WEEKOFYEAR][PadIndex],
							pt, ptlim);
					else if (*format == 'g') {
						pt = _yconv(year, base, 0, 1,
							pt, ptlim);
					} else	pt = _yconv(year, base, 1, 1,
							pt, ptlim);
				}
				continue;
			case 'v':
				/*
				 * From Arnold Robbins' strftime version 3.0:
				 * "date as dd-bbb-YYYY"
				 * (ado, 1993-05-24)
				 */
				pt = _fmt("%e-%b-%Y", t, pt, ptlim);
				continue;
			case 'W':
				pt = _conv((t->tm_yday + DAYSPERWEEK -
					(t->tm_wday ?
					(t->tm_wday - 1) :
					(DAYSPERWEEK - 1))) / DAYSPERWEEK,
					fmt_padding[PAD_FMT_WEEKOFYEAR][PadIndex],
					pt, ptlim);
				continue;
			case 'w':
				pt = _conv(t->tm_wday, "%d", pt, ptlim);
				continue;
			case 'X':
				pt = _fmt(tptr->X_fmt, t, pt, ptlim);
				continue;
			case 'x':
				pt = _fmt(tptr->x_fmt, t, pt, ptlim);
				continue;
			case 'y':
				pt = _yconv(t->tm_year, TM_YEAR_BASE, 0, 1,
					pt, ptlim);
				continue;
			case 'Y':
				pt = _yconv(t->tm_year, TM_YEAR_BASE, 1, 1,
					pt, ptlim);
				continue;
			case 'Z':
#ifdef TM_ZONE
				if (t->TM_ZONE != NULL)
					pt = _add(t->TM_ZONE, pt, ptlim);
				else
#endif /* defined TM_ZONE */
				if (t->tm_isdst >= 0)
					pt = _add(tzname[t->tm_isdst != 0],
						pt, ptlim);
				/*
				 * C99 says that %Z must be replaced by the
				 * empty string if the time zone is not
				 * determinable.
				 */
				continue;
			case 'z':
				{
				int		diff;
				char const *	sign;

				if (t->tm_isdst < 0)
					continue;
#ifdef TM_GMTOFF
				diff = t->TM_GMTOFF;
#else /* !defined TM_GMTOFF */
				/*
				 * C99 says that the UTC offset must
				 * be computed by looking only at
				 * tm_isdst. This requirement is
				 * incorrect, since it means the code
				 * must rely on magic (in this case
				 * altzone and timezone), and the
				 * magic might not have the correct
				 * offset. Doing things correctly is
				 * tricky and requires disobeying C99;
				 * see GNU C strftime for details.
				 * For now, punt and conform to the
				 * standard, even though it's incorrect.
				 *
				 * C99 says that %z must be replaced by the
				 * empty string if the time zone is not
				 * determinable, so output nothing if the
				 * appropriate variables are not available.
				 */
				if (t->tm_isdst == 0)
#ifdef USG_COMPAT
					diff = -timezone;
#else /* !defined USG_COMPAT */
					continue;
#endif /* !defined USG_COMPAT */
				else
#ifdef ALTZONE
					diff = -altzone;
#else /* !defined ALTZONE */
					continue;
#endif /* !defined ALTZONE */
#endif /* !defined TM_GMTOFF */
				if (diff < 0) {
					sign = "-";
					diff = -diff;
				} else
					sign = "+";
				pt = _add(sign, pt, ptlim);
				diff /= SECSPERMIN;
				diff = (diff / MINSPERHOUR) * 100 +
					(diff % MINSPERHOUR);
				pt = _conv(diff,
					fmt_padding[PAD_FMT_YEAR][PadIndex],
					pt, ptlim);
				}
				continue;
			case '+':
				pt = _fmt(tptr->date_fmt, t, pt, ptlim);
				continue;
			case '-':
				if (PadIndex != PAD_DEFAULT)
					break;
				PadIndex = PAD_LESS;
				goto label;
			case '_':
				if (PadIndex != PAD_DEFAULT)
					break;
				PadIndex = PAD_SPACE;
				goto label;
			case '0':
				if (PadIndex != PAD_DEFAULT)
					break;
				PadIndex = PAD_ZERO;
				goto label;
			case '%':
			/*
			 * X311J/88-090 (4.12.3.5): if conversion char is
			 * undefined, behavior is undefined. Print out the
			 * character itself as printf(3) also does.
			 */
			default:
				break;
			}
		}
		if (pt == ptlim)
			break;
		*pt++ = *format;
	}
	return pt;
}

static char *
_conv(int n, const char *format, char *pt,
	const char *ptlim)
{
	char	buf[INT_STRLEN_MAXIMUM(int) + 1];

	(void) sprintf(buf, format, n);
	return _add(buf, pt, ptlim);
}

static char *
_secs(const struct tm *t, char *pt, const char *ptlim)
{
	struct tm	tm;
	char		buf[INT_STRLEN_MAXIMUM(time_t) + 1];
	time_t		mkt;

	/* Make a copy, mktime(3) modifies the tm struct. */
#ifdef __ORCAC__
	/*
	 * ORCA/C pukes with an "illegal operand for the indirection
	 * operator" message
	 */
	memcpy(&tm, t, sizeof(struct tm));
#else
	tm = *t;
#endif
	mkt = mktime(&tm);
	(void) sprintf(buf, "%ld", (long)mkt);
	return _add(buf, pt, ptlim);
}
  
static char *
_add(const char *str, char *pt, const char *ptlim)
{
	while (pt < ptlim && (*pt = *str++) != '\0')
		++pt;
	return pt;
}

/*
 * POSIX and the C Standard are unclear or inconsistent about
 * what %C and %y do if the year is negative or exceeds 9999.
 * Use the convention that %C concatenated with %y yields the
 * same output as %Y, and that %Y contains at least 4 bytes,
 * with more only if necessary.
 */

static char *
_yconv(int a, int b, int convert_top, int convert_yy,
    char *pt, const char *ptlim)
{
	register int	lead;
	register int	trail;

#define	DIVISOR	100
	trail = a % DIVISOR + b % DIVISOR;
	lead = a / DIVISOR + b / DIVISOR + trail / DIVISOR;
	trail %= DIVISOR;
	if (trail < 0 && lead > 0) {
		trail += DIVISOR;
		--lead;
	} else if (lead < 0 && trail > 0) {
		trail -= DIVISOR;
		++lead;
	}
	if (convert_top) {
		if (lead == 0 && trail < 0)
			pt = _add("-0", pt, ptlim);
		else	pt = _conv(lead, "%02d", pt, ptlim);
	}
	if (convert_yy)
		pt = _conv(((trail < 0) ? -trail : trail), "%02d", pt,
		     ptlim);
	return (pt);
}
